http.go

```diff
@@ -1360,7 +1363,9 @@ func (req *Request) Write(w *bufio.Writer) error {
 		if len(host) == 0 {
 			return errRequestHostRequired
 		}
-		req.Header.SetHostBytes(host)
+		if len(req.Header.Host()) == 0 || !req.AllowToChangeHostHeader {
+			req.Header.SetHostBytes(host)
```

Comment:
Shouldn't the `if len(host) == 0 { return errRequestHostRequired` check then also be inside the `if req.AllowToChangeHostHeader`?

Reply:
I don't think so. The uri.Host() used as connection address at least for Client type. If host will be empty but Host header is not, connection will be broken. I've checked such code in http.go

		if len(req.Header.Host()) == 0 || !req.UseHostHeader {
			req.Header.SetHostBytes(host)
		} else if len(host) == 0 {
			return errRequestHostRequired
		}

And such usage for this code

		req.UseHostHeader = true
		req.SetHost("")
		req.Header.SetHost(host)

There is result - lookup : no such host

We can use headers for setting uri.Host() but I am not sure that it is correct.

Reply:
I've check code again. It is not important for Client, because error raised inside dialAddr. But I am still not sure that it is correct allow uri.Host() to be empty if Host header presents.

Reply:
But shouldn't `len(host) == 0` be allowed when `req.UseHostHeader === true`? It makes more sense to me to put that inside the if?

_Sorry I'm replying bit slow, I'm traveling._

Reply:
I've checked logic again and pushed a new version. errRequestHostRequired should be returned only if both uri.Host() and req.Header.Host() are empty. And it does not depend from UseHostHeader. If Header.Host() is empty we always need to set header from uri.Host(). And only when both are not empty we need to check UseHostHeader.